event-protocol/README.md

```diff
@@ -0,0 +1,150 @@
+# Cucumber Event Protocol
+
+The Cucumber event protocol is a data exchange format for components in the
+Cucumber ecosystem.
+
+When Cucumber runs, it produces (emits) Cucumber events that are consumed by
+formatters that report the results in various formats. The protocol is designed
+to allow for other producers to emit events, and for other consumers to consume
+them.
+
+The protocol is transport agnostic, and can be transmitted in-process as well as
+between processes. This means some components (such as the [HTML Formatter](../html-formatter/README.md))
+can be implemented in a single programming language and used by Cucumber
+implementations in different languages.
+
+The Cucumber event protocol can also be used to implement parallel and distributed
+execution of Cucumber scenarios.
+
+Programs that support the protocol can be producers or consumers (or both).
+
+Examples of producers are:
+
+* Cucumber
+* [Gherkin-Lint](../gherkin-lint/README.md)
+
+Examples of consumers are:
+* Formatters (HTML, JUnit, Progress, PDF)
+* 3rd party tools (Cucumber Pro)
+
+## Events {#events}
+
+### start {#event-start}
+
+Example:
+[import](examples/events/001_start.json)
+
+### source {#event-source}
+
+Example:
+[import](examples/events/002_source.json)
+
+### attachment {#event-attachment}
+
+An `attachment` attaches a piece of metadata to a particular line in a `source` file.
+Attachments can have many types (specified by a media type), and are typically used for:
+
+* `text/vnd.cucumber.step.status+plain`, status of a failing step (passed, undefined, pending, failed, skipped)
+* `text/vnd.cucumber.stacktrace.java+plain`, stack trace of a failing step (from java in this case)
+* `image/png`, `image/jpg`, `image/gif` - screenshots
+
+Example:
+[import](examples/events/003_attachment-png-embedded.json)
+
+## Implementation
+
+* Cucumber events are formatted as [Newline Delimited JSON](http://ndjson.org)
+* The transport protocol can use one or more of the following:
+  * `STDIN`/`STDOUT`
+  * Raw socket
+  * W3C WebSocket
+  * W3C EventSource (each event is a JSON-encoded event without `event:` type - it's in the payload itself)
+
+### Philosophy
+
+As the event specification evolves to support a richer set of events there are some
+key principles to consider:
+
+#### File format agnostic
+
+While Gherkin is currently the only file format that will be used in `source`
+events, no events should assume that all files will be Gherkin documents. This
+allows for alternative document formats in the future.
+
+#### No Cucumber execution semantics
+
+Some consumers (such as the [HTML Formatter](../html-formatter/README.md)) may
+consume events emitted by other consumers than Cucumber (for example [Gherkin-Lint](../gherkin-lint/README.md)).
+
+For this reason, events containing information about execution should be represented
+in a more generic way.
```

Comment:
This is true. Any linter will necessarily have to interact at the AST level because a rule like 'you forget to add any steps to this placeholder scenario' will not be doable because Pickles would not be generated for such a scenario.